tools/kprint/addBrackets.cpp

```diff
@@ -0,0 +1,202 @@
+      return true;
+    }
+    if (data.terminals.at(innerName) == "0") {
+      return false;
+    }
+    Fixity innerFixity = getFixity(innerComposite, data);
+    if ((innerFixity & BARE_RIGHT) && rightCapture != nullptr) {
+      bool inversePriority = isPriorityWrong(innerComposite, rightCapture, innerComposite->getArguments().size() - 1, data);
+      Fixity rightCaptureFixity = getFixity(rightCapture, data);
+      if (!inversePriority && (rightCaptureFixity & BARE_LEFT)) {
+        return true;
+      }
+    }
+    if ((innerFixity & BARE_LEFT) && leftCapture != nullptr) {
+      bool inversePriority = isPriorityWrong(innerComposite, leftCapture, 0, data);
+      Fixity leftCaptureFixity = getFixity(leftCapture, data);
+      if (!inversePriority && (leftCaptureFixity & BARE_RIGHT)) {
+        return true;
+      }
+    }
+    return false;
+  } else {
+    return false;
+  }
+}
+
+sptr<KOREPattern> addBrackets(sptr<KOREPattern> inner, KORECompositePattern *outer, KORECompositePattern *leftCapture, KORECompositePattern *rightCapture, int position, PrettyPrintData const& data) {
+  if (auto innerComposite = dynamic_cast<KORECompositePattern *>(inner.get())) {
+    if (innerComposite->getConstructor()->getName() == "inj") {
+      return addBrackets(innerComposite->getArguments()[0], outer, leftCapture, rightCapture, position, data);
+    }
+  }
+  if (requiresBracketWithSimpleAlgorithm(outer, leftCapture, rightCapture, inner.get(), position, data)) {
+    sptr<KORESort> outerSort = outer->getConstructor()->getArguments()[position];
+    sptr<KORESort> innerSort = inner->getSort();
+    for (auto &entry : data.brackets) {
+      bool isCorrectOuterSort = lessThanEq(data, entry.first, outerSort.get());
+      if (isCorrectOuterSort) {
+        for (KORESymbol *s : entry.second) {
+          bool isCorrectInnerSort = lessThanEq(data, innerSort.get(), s->getArguments()[0].get());
+          if (isCorrectInnerSort) {
+            sptr<KORECompositePattern> result = KORECompositePattern::Create(s);
+            result->addArgument(inner);
+            return result;
+          }
+        }
+      }
+    }
+    sptr<KORECompositePattern> result = KORECompositePattern::Create("bracket");
+    result->addArgument(inner);
+    result->getConstructor()->addSort(innerSort);
+    return result;
+  }
+  return inner;
+}
+
+sptr<KOREPattern> addBrackets(sptr<KOREPattern> t, KORECompositePattern *previousLeftCapture, KORECompositePattern *previousRightCapture, PrettyPrintData const& data) {
+  if (auto outer = dynamic_cast<KORECompositePattern *>(t.get())) {
+    if (outer->getConstructor()->getName() == "\\dv") {
+      return t;
+    }
+    std::vector<sptr<KOREPattern>> newItems;
+
+    sptr<KORECompositePattern> result = KORECompositePattern::Create(outer->getConstructor());
+    int position = 0;
+    for (auto &inner : outer->getArguments()) {
+      KORECompositePattern *leftCapture = getLeftCapture(previousLeftCapture, outer, position, data);
```

Comment:
I was trying to figure these things out, but it's fairly slow, so I thought it's better to ask.

What do `BARE_LEFT` and `BARE_RIGHT` mean? What does `capture` in `leftCapture` and `rightCapture` mean?

Is it worth adding code comments to explain some of these, or it's just me not having the context/knowledge that other people that work with this code have?

Is the `terminals` attribute documented anywhere? If yes, where? I tried searching for it in the `pending-documentation.md` file, but I didn't find it. If not, should it be?

Reply:
The terminals attribute is one of the attributes that gets automatically inserted into the kore file by ModuleToKORE in the frontend. It exists because the kore data otherwise does not track the relative position of terminals and nonterminals in the original production, which makes it impossible to correctly compute where to add brackets according to this algorithm. I will document the algorithm in more detail in this file. Mostly it's a straight translation of an algorithm that exists in the K frontend which is in turn a straight translation of an algorithm that exists in the maude codebase, where it definitely wasn't documented because maude's code has basically no documentation.